crates/storage/provider/src/providers/state/historical.rs

```diff
@@ -166,48 +176,36 @@ impl<'b, Provider: DBProvider + BlockNumReader> HistoricalStateProviderRef<'b, P
     {
         let mut cursor = self.tx().cursor_read::<T>()?;
 
-        // Lookup the history chunk in the history index. If they key does not appear in the
+        // Lookup the history chunk in the history index. If the key does not appear in the
         // index, the first chunk for the next key will be returned so we filter out chunks that
         // have a different key.
-        if let Some(chunk) = cursor.seek(key)?.filter(|(key, _)| key_filter(key)).map(|x| x.1 .0) {
+        if let Some(chunk) = cursor.seek(key)?.filter(|(key, _)| key_filter(key)).map(|x| x.1) {
             // Get the rank of the first entry before or equal to our block.
             let mut rank = chunk.rank(self.block_number);
 
             // Adjust the rank, so that we have the rank of the first entry strictly before our
             // block (not equal to it).
-            if rank.checked_sub(1).and_then(|rank| chunk.select(rank)) == Some(self.block_number) {
-                rank -= 1
-            };
+            if rank.checked_sub(1).and_then(|r| chunk.select(r)) == Some(self.block_number) {
+                rank -= 1;
+            }
 
-            let block_number = chunk.select(rank);
+            let found_block = chunk.select(rank);
 
             // If our block is before the first entry in the index chunk and this first entry
             // doesn't equal to our block, it might be before the first write ever. To check, we
             // look at the previous entry and check if the key is the same.
             // This check is worth it, the `cursor.prev()` check is rarely triggered (the if will
             // short-circuit) and when it passes we save a full seek into the changeset/plain state
             // table.
-            if rank == 0 &&
-                block_number != Some(self.block_number) &&
```

Comment:
these 2 logic goes into `needs_prev_shard_check` imo its easier to read and understand 

the part i didnt like about the previous code was that the code has too many conditionals and was hard to hold in my head